3-enrich/scala-common-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/registry/AnonIpEnrichment.scala

```diff
@@ -77,64 +88,144 @@ object AnonOctets extends Enumeration {
+      case _: NoSuchElementException =>
+        s"IPv6 address segments to anonymize must be 1, 2, 3, 4, 5, 6, 7 or 8. Value $anonIPv6Segments was given".toProcessingMessage.fail
     }
 }
 
 /**
  * Config for an anon_ip enrichment
  *
- * @param octets The number of octets to anonymize
+ * @param ipv4Octets The number of octets (IPv4) to anonymize
+ * @param ipv6Segments The number of segments (IPv6) to anonymize
  */
 case class AnonIpEnrichment(
-  octets: AnonOctets.AnonOctets
+  ipv4Octets: AnonIPv4Octets.AnonIPv4Octets,
+  ipv6Segments: AnonIPv6Segments.AnonIPv6Segments
 ) extends Enrichment {
 
+  val IPv4MappedAddressPrefix = "::FFFF:"
+  val MaskChar                = "x"
+
   /**
    * Anonymize the supplied IP address.
    *
-   * octets is the number of octets
-   * in the IP address to anonymize, starting
-   * from the right. For example:
+   * ipv4Octets is the number of octets in the IPv4 address to anonymize, starting
+   * from the right (example below)
+   *
+   * ipv6Segments is the number of segments in the IPv6 address to anonymize, starting
+   * from the right (exaple below)
    *
-   * anonymizeIp("94.15.223.151", One)
-   * => "94.15.223.x"
+   * example:
    *
-   * anonymizeIp("94.15.223.151", Three)
+   * val enrichment = AnonIpEnrichment(Three, Four)
+   * enrichment.anonymizeIp("94.15.223.151")
    * => "94.x.x.x"
    *
+   * enrichment.anonymizeIp("2605:2700:0:3::4713:93e3")
+   * => "2605:2700:0:3:x:x:x:x"
+   *
+   * For IPv6 either the form defined in RFC 2732
+   *  or the literal IPv6 address format defined in RFC 2373 is accepted
+   *
    * TODO: potentially update this to return
    * a Validation error or a null if the IP
    * address is somehow invalid or incomplete.
    *
    * @param ip The IP address to anonymize
    * @return the anonymized IP address
    */
-  import AnonOctets._
-  def anonymizeIp(ip: String): String =
-    Option(ip)
-      .map(
-        _.split("\\.").zipWithIndex
-          .map {
-            case (q, i) => {
-              if (octets.id >= All.id - i) "x" else q
-            }
-          }
-          .mkString("."))
-      .orNull
+  def anonymizeIp(ipOrNull: String): String =
+    Option(ipOrNull).map { ip =>
+      Try(GuavaInetAddress.forString(ip))
+        .map {
+          case _: Inet4Address    => anonymizeIpV4(ip)
+          case ipv6: Inet6Address => anonymizeIpV6(ipv6.getHostAddress)
+        }
+        .getOrElse(tryAnonymizingInvalidIp(ip))
+    }.orNull
+
+  private def anonymizeIpV4(ip: String): String = {
+    def masked(ipv4: String) = {
+      val split = ipv4.split("\\.")
+      split
+        .take(AnonIPv4Octets.All.id - ipv4Octets.id)
+        .toList
+        .padTo(split.size, MaskChar)
+        .mkString(".")
+    }
+
+    if (ip.startsWith(IPv4MappedAddressPrefix))
+      IPv4MappedAddressPrefix + masked(ip.replace(IPv4MappedAddressPrefix, ""))
+    else masked(ip)
+  }
+
+  private def anonymizeIpV6(ip: String): String =
```

Comment:
`anonymizeIpV4` and `anonymizeIpV6` could be combined into one generic function

Reply:
Theoretically it could be (I was also thinking about it ;)), but the generic version would need to take 4 params, so I don't think it would make it better actually.